idp/openid/openid-connect.go

```diff
@@ -223,50 +248,72 @@ func (idp *openidConnectIdentityProvider) callback(ctx context.Context, w http.R
 	if err != nil {
 		return errgo.Mask(err)
 	}
-	idtok := tok.Extra("id_token")
-	if idtok == nil {
-		return errgo.Newf("no id_token in OpenID response")
-	}
-	idtoks, ok := idtok.(string)
-	if !ok {
-		return errgo.Newf("invalid id_token in OpenID response")
+
+	ic := idp.params.IdentityCreator
+	if ic == nil {
+		ic = idp
 	}
-	id, err := idp.provider.Verifier(&oidc.Config{ClientID: idp.config.ClientID}).Verify(ctx, idtoks)
+	user, err := ic.CreateIdentity(ctx, tok)
 	if err != nil {
 		return errgo.Mask(err)
 	}
-	user := store.Identity{
-		ProviderID: store.MakeProviderIdentity(idp.Name(), fmt.Sprintf("%s:%s", id.Issuer, id.Subject)),
+	
+	existingUser := store.Identity{
+		ProviderID: user.ProviderID,
 	}
-	err = idp.initParams.Store.Identity(ctx, &user)
+	err = idp.initParams.Store.Identity(ctx, &existingUser)
 	if err == nil {
-		idp.initParams.VisitCompleter.RedirectSuccess(ctx, w, req, ls.ReturnTo, ls.State, &user)
-		return nil
+		var upd store.Update
+		// A user exists check if it needs updating.
+		if user.Name != "" && existingUser.Name != user.Name {
+			existingUser.Name = user.Name
+			upd[store.Name] = store.Set
+		}
+		if user.Email != "" && existingUser.Email != user.Email {
+			existingUser.Email = user.Email
+			upd[store.Email] = store.Set
+		}
+		if (upd != store.Update{}) {
+			err = idp.initParams.Store.UpdateIdentity(ctx, &existingUser, upd)
+		}
+		if err == nil {
```

Comment:
nitpick: this is a bit confusing as the only way to the err to be not nil comes from line 277..

Reply:
yes but line #277 might not have been attempted, and if it was, and failed, we want to carry on with the function.